File: js/stage3/S3_PIPE_06.js
// S3_PIPE_06 — Safe pipeline
const pipeSafe =
  (...fns) =>
  (value) => {
    try {
      let result = value;
      for (const fn of fns) {
        result = fn(result);
      }
      return { ok: true, value: result };
    } catch (error) {
      return { ok: false, error: error.message };
    }
  };
// Tests
const safeFn = pipeSafe(
  (x) => x + 1,
  (x) => {
    if (x > 3) throw new Error("Too big");
    return x;
  }
);
console.log(safeFn(1)); // ok
console.log(safeFn(5)); // error

File: Python/stage3/S3_PIPE_06.py
# S3_PIPE_06 — Safe pipeline with error handling

def pipe_safe(*fns):

    def runner(x):

        try:

            for fn in fns:

                x = fn(x)

            return {"ok": True, "value": x}

        except Exception as e:

            return {"ok": False, "error": str(e)}

    return runner


# Functions

def to_int(x):

    return int(x)

def invert(x):

    return 1 / x


safe_pipeline = pipe_safe(to_int, invert)

print(safe_pipeline("5"))

print(safe_pipeline("0"))

print(safe_pipeline("abc"))
 